### PR TITLE
Add progress bar when downloading python

### LINKED ISSUE #4825: Python downloads should show a progress bar
When downloading a python toolchain for running a tool, i wait several seconds without any output. This is very untypical and feels wrong for uv. We should add a message indicating the start of a large download and a progress bar. The message should be shown anyway since non-interactive environments don't show progress bars. (The video is from the failing https://github.com/astral-sh/uv/pull/4810, but it applies also when this is fixed)

[Screencast from 2024-07-05 10-01-24.webm](https://github.com/astral-sh/uv/assets/6826232/19a39288-4f78-426a-8d9a-e9a5653c242e)

### PR DESCRIPTION
**Pull request by j178**
## Summary

Resolves #4825 

## Test Plan

```sh
$ cargo run -- python install --force --preview
$ cargo run -- venv -p 3.12 --python-preference only-managed
$ cargo run -- tool install --preview -p 3.12 --python-preference only-managed --force black
````

### CODE REVIEW

File: crates/uv-python/src/downloads.rs [hunk 1]
```diff
@@ -420,46 +424,63 @@ impl ManagedPythonDownload {
...
             .bytes_stream()
             .map_err(|err| io::Error::new(io::ErrorKind::Other, err))
             .into_async_read();
 
+        let mut hashers = self
+            .sha256
+            .into_iter()
+            .map(|_| Hasher::from(HashAlgorithm::Sha256))
+            .collect::<Vec<_>>();
+        let mut hasher = uv_extract::hash::HashReader::new(stream.compat(), &mut hashers);
```
Comment (charliermarsh) [inline crates/uv-python/src/downloads.rs:450]: I was kinda hoping to avoid hashing when `self.sha256` is `None`, any chance we can preserve that behavior?

**Comment by charliermarsh** [comment]
Thanks for doing this, awesome!

File: crates/uv-python/src/downloads.rs (hunk 1 above)
Comment (j178) [reply crates/uv-python/src/downloads.rs:450]: If `self.sha256` is `None`, `hashers` will be empty, and `HashReader` won't do any hashing. So I think the overhead might be negligible. 
https://github.com/astral-sh/uv/blob/1bd73a7346ca62e3a7440f32980fa59591535610/crates/uv-extract/src/hash.rs#L138-L140

**Comment by charliermarsh** [comment]
Can we try to omit the `░░░░░░░░░░░░░░░░ [1/1] Downloading Python...` when we're downloading exactly one version?

**Review by charliermarsh — APPROVED** [review]

**Pull request merged**